### ISSUE TITLE
Jaeger Ingest for Elasticsearch 7.0 release date

### ISSUE THREAD

**Issue by yichingwu** [jaegertracing/jaeger#1474]
Since 7.0 just came out of Elasticsearch, how long will it typically take for Jaeger Ingest to be updated to 7.0?

**Comment by eric-price** [comment]
Elasticsearch 7.0 breaks Jaeger for me. Getting the following:

`Apr 17 03:59:11 eng2 jaeger-collector[19190]: {"level":"error","ts":1555473551.6439335,"caller":"spanstore/writer.go:190","msg":"Failed to create index","trace_id":"77d523381c0350b","span_id":"77d523381c0350b","error":"elastic: Error 400 (Bad Request): Setting index.mapper.dynamic was removed after version 6.0.0 [type=illegal_argument_exception]","stacktrace":"github.com/jaegertracing/jaeger/plugin/storage/es/spanstore.(*SpanWriter).logError\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/plugin/storage/es/spanstore/writer.go:190\ngithub.com/jaegertracing/jaeger/plugin/storage/es/spanstore.(*SpanWriter).createIndex\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/plugin/storage/es/spanstore/writer.go:154\ngithub.com/jaegertracing/jaeger/plugin/storage/es/spanstore.(*SpanWriter).WriteSpan\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/plugin/storage/es/spanstore/writer.go:117\ngithub.com/jaegertracing/jaeger/cmd/collector/app.(*spanProcessor).saveSpan\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/collector/app/span_processor.go:101\ngithub.com/jaegertracing/jaeger/cmd/collector/app.(*spanProcessor).saveSpan-fm\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/collector/app/span_processor.go:88\ngithub.com/jaegertracing/jaeger/cmd/collector/app.ChainedProcessSpan.func1\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/collector/app/model_consumer.go:34\ngithub.com/jaegertracing/jaeger/cmd/collector/app.(*spanProcessor).processItemFromQueue\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/collector/app/span_processor.go:127\ngithub.com/jaegertracing/jaeger/cmd/collector/app.NewSpanProcessor.func1\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/collector/app/span_processor.go:56\ngithub.com/jaegertracing/jaeger/pkg/queue.(*BoundedQueue).StartConsumers.func1\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/pkg/queue/bounded_queue.go:65"}`

**Comment by pavolloffay** [comment]
It might take some time. We will need to update ES go client and probably change some mappings. At the same time we want to keep compatibility with older ES versions so we need to figure out how to do it.

**Comment by pavolloffay** [comment]
A related issue is also https://github.com/jaegertracing/jaeger/issues/1406

**Comment by Asgoret** [comment]
Any information?)

**Comment by yichingwu** [comment]
is there any update for this issue ?

**Comment by yurishkuro** [comment]
To my knowledge, nobody is working on this. It's up for grabs, hence the label.

**Comment by CpuID** [comment]
Is this just simply a matter of adjusting the upstream index template stuff?

a la

https://github.com/jaegertracing/jaeger/blob/master/plugin/storage/es/mappings/jaeger-service.json#L8

https://github.com/jaegertracing/jaeger/blob/master/plugin/storage/es/mappings/jaeger-span.json#L8

**Comment by pavolloffay** [comment]
It will require bumping the ES client version and then make the intext templates compatible with ES7. 

We should also consider migration to https://github.com/elastic/go-elasticsearch instead of currently used https://github.com/olivere/elastic

**Comment by yichingwu** [comment]
@pavolloffay ,it's a  good news for us, thanks a lot.

**Comment by magicdvd** [comment]
@pavolloffay is there any further information for  es7 support?

**Comment by wangycc** [comment]
which version will support es7?

**Comment by Nilubkal** [comment]
Hey guys, do you have any additional info when elk v7.x support will be introduced ?

**Comment by pavolloffay** [comment]
There isn't any news. Any updates will be posted to this thread.

**Comment by gregoryfranklin** [comment]
I had a bit of a look at this to see whether I could put together a pull request.

I tested against elasticsearch 5.6.12, 6.8.1 and 7.2.0

Upgrading github.com/olivere/elastic to 7.0.4 was straight forward.

Attempting to switch to github.com/elastic/go-elasticsearch ran into problems because 'dep' has problems with go modules when using v7 as a major version.  It would appear that either 'dep' needs to be fixed (there is an open issue for this) or jaeger needs to move to go modules to use github.com/elastic/go-elasticsearch.

Elasticsearch 7.x required a few changes.  It no longer supports Document Types, the '_all' field or the '_default_' mapping.  Also 'track_total_hits' needed to be enabled.
After making these changes, the integration test ran successfully against 7.2.0.  Unfortunately, the test no longer worked against 6.8.1 or 5.6.12.  So, further changes will be needed to support multiple versions of Elasticsearch, either by enabling deprecated features in elasticsearch or doing version detection in jaeger.

It would probably make it easier if support for elasticsearch 5.x was dropped, it went end-of-life on 2019-03-11

I'll see if I can get it working against both 6.x and 7.x using github.com/olivere/elastic at which point I will post some code (assuming I find the time).

**Comment by gregoryfranklin** [comment]
I had to request a change to github.com/olivere/elastic to support both 6.x and 7.x

https://github.com/olivere/elastic/pull/1146

If / when this is merged, I will have a working pull request that passes the integration tests against elasticsearch 6.8.1 and 7.2.0

**Comment by pavolloffay** [comment]
great update @gregoryfranklin, thanks for looking into this!

Ultimately, I wanted to keep the compatibility with ES5 for legacy users and just include a new flag or storage type for ES7. Let's see how the PR looks.

We will also merge  https://github.com/jaegertracing/jaeger/pull/1627 which changes from index creating to index templates. As the templates are in separate files we could change those for appropriate ES version at boot time.

**Comment by gregoryfranklin** [comment]
I created the PR (#1690) so you can see what it looks like.

ES5 support could probably be preserved with a couple more tweaks.  I'd have to run some more tests.

However, to support multiple ES versions, you are really looking at needing to switch Jaeger to go modules so that you can use different major versions of github.com/elastic/go-elasticsearch concurrently.  You would then either use a config option or query the elasticsearch server to determine the version and use a different client instance based on that version.

Has any thought been put into to switching from 'dep' to go modules?

**Comment by gregoryfranklin** [comment]
I had a bit of a look at what would be required to support the version specific client libraries (since both github.com/olivere/elastic and github.com/elastic/go-elasticsearch only support specific versions of the library against their matching ES version)

None of the v7 clients work with 'dep' - they require go modules.  However, it was possible to use gopkg.in/olivere/elastic.v5 and github.com/olivere/elastic (v6) concurrently.

For writes (CreateIndex, Index etc), it was relatively straight forward.  All the version specific logic can be extracted into the ClientWrapper.

Disentangling the reads (Search, MultiSearch) is substantially more difficult and I did not attempt it.

The following is approximately what it might look like.  Its quite messy, but code readability could definitely be improved.
https://github.com/gregoryfranklin/jaeger/commit/438ab3d354f1a42dcd96a81499b8e8b82eb21550

This could be considered as part of a longer term re-factor (moving to github.com/elastic/go-elasticsearc and go modules), but I think we can make support for ES5, ES6 and ES7 work using github.com/olivere/elastic (v6).

**Comment by pavolloffay** [comment]
@gregoryfranklin what is the current blocker for using only olivere library? Only https://github.com/olivere/elastic/pull/1146? 

Or the third solution might be to implement our own client and talk to ES through it. It seems pretty easy, the complication is the bulk processor.

I was also thinking that we could automatically adjust mappings/clients based on the ES version returned from rest endpoint.

**Comment by gregoryfranklin** [comment]
I've updated the PR.  The storage-integration-test passes against 5.6.12, 6.8.1 and 7.2.0 (assuming https://github.com/olivere/elastic/pull/1146) is merged.

I've just noticed I also need to look at esRollover.py

**Comment by pavolloffay** [comment]
Great work @gregoryfranklin! I will wait until the https://github.com/olivere/elastic/pull/1146 is merged and then have a look.

**Comment by GuardMoony** [comment]
@pavolloffay can you check if everything is working now? olivere/elastic#1146 issue has been fixed and released.

**Comment by pavolloffay** [comment]
Yes I am reviewing #1690

**Comment by GuardMoony** [comment]
thx for the update

**Comment by pavolloffay** [comment]
@jaegertracing/elasticsearch the `latest` images now support ES7. Enjoy! And thanks to @gregoryfranklin!!!

**Comment by kiddo3** [comment]
Any way to make it work with jaeger-operator? Bacause the image used is 1.13.1.

**Comment by pavolloffay** [comment]
Yes, you can override collector and query images to `latest` in the CR.

**Comment by doctorpangloss** [comment]
@pavolloffay I'm using the latest images, but I still get:

```
HTTP Error: Search service failed: json: cannot unmarshal object into Go struct field SearchHits.total of type int64
```

**Comment by pavolloffay** [comment]
@doctorpangloss what is the issue? I do not understand you? 

The `latest` jaeger images do not work with ES7 while 1.14.0 work? Could you please be more specific?

**Comment by doctorpangloss** [comment]
Yes, the image hash I specified, the current hash for `latest`, broke 7.x support while 1.14 does work.

Reading this thread in GitHub, we’re instructed to use `latest`, and I think I was just unlucky enough to use it on the day it breaks.

**Comment by doctorpangloss** [comment]
@pavolloffay Just to follow up, it looks like https://hub.docker.com/u/wjang deleted that particular image, because it was almost certainly built wrong, from docker, and then rebuilt latest, but I wouldn't be able to tell you if it works.

**Comment by pavolloffay** [comment]
We run tests against ES 5.x, 6.x and 7.x for every build on Jaeger so if something is failing it is most likely related to a wrong configuration or using non-standard ES images.

**Comment by CoderPoet** [comment]
It seems that the go es client currently used does not support [index template api](https://www.elastic.co/guide/en/elasticsearch/reference/7.13/index-templates.html)
